contracts/mainnet/connectors/euler/main.sol

```diff
@@ -0,0 +1,552 @@
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		eToken.mint(subAccount, amt);
+
+		setUint(setId, _amt);
+
+		_eventName = "LogMint(uint256,address,uint256,uint256,uint256)";
+		_eventParam = abi.encode(subAccount, token, _amt, getId, setId);
+	}
+
+	/**
+	 * @dev Burn ETH/ERC20_Token.
+	 * @notice Burn a token from Euler. Burn removes equal amount of deposits and debts.
+	 * @param subAccount Subaccount number
+	 * @param token The address of the token to burn.(For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param amt The amount of the token to burn. (For max: `uint256(-1)`)
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of tokens deposited.
+	 */
+	function burn(
+		uint256 subAccount,
+		address token,
+		uint256 amt,
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+
+		IEulerDToken dToken = IEulerDToken(markets.underlyingToDToken(_token));
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		_amt = _amt == type(uint256).max
+			? dToken.balanceOf(address(this))
+			: _amt;
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		eToken.burn(subAccount, amt);
+
+		setUint(setId, _amt);
+
+		_eventName = "LogBurn(uint256,address,uint256,uint256,uint256)";
+		_eventParam = abi.encode(subAccount, token, _amt, getId, setId);
+	}
+
+	/**
+	 * @dev ETransfer ETH/ERC20_Token.
+	 * @notice ETransfer deposits from account to another.
+	 * @param subAccountFrom subAccount from which deposit is transferred
+	 * @param subAccountTo subAccount to which deposit is transferred
+	 * @param token The address of the token to etransfer.(For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param amt The amount of the token to etransfer. (For max: `uint256(-1)`)
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of tokens deposited.
+	 */
+	function eTransfer(
+		uint256 subAccountFrom,
+		uint256 subAccountTo,
+		address token,
+		uint256 amt,
+		uint256 getId,
+		uint256 setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		_amt = _amt == type(uint256).max
+			? eToken.balanceOf(address(this))
+			: _amt;
+
```

Comment:
eTokens are interest-bearing tokens, so it increases over time. Better to use `balanceOfUnderlying` if the user wants to transfer the whole eTokens amount, including interest accumulated.

Reply:
https://github.com/euler-xyz/euler-contracts/blob/master/contracts/modules/EToken.sol#L79

Reply:
`balanceOfUnderlying` returns the balance in token decimals, while `balanceOf` returns it in its own internal unit.

Reply:
I think we should confirm this difference with the Euler team?

Reply:
Resolved!